Enable clients to enforce an ACR via client attribute

sonOfRa opened the pull request:
closes #16884 

This is a possible implementation of https://github.com/keycloak/keycloak/issues/16884 using a client attribute in order to enforce a specific ACR on a client.

This enables users to enforce 2FA via ACR even when the client application does not support scanning the received tokens for the desired ACR level

sonOfRa commented:
Hi @thomasdarimont, thanks for the feedback!

Q1) This generally seems like a good idea, but it seems hard to implement correctly: while LoA is numeric and "higher" means "better", ACR are just arbitrary parameters, mapped to some LoA values. Now, how do we correctly interpret whether an acr_values parameter supplied by the client (or a malicious user!) still satisfies the minimum requirement we want for the client? Simply merging the sets of minimal acr_values with the client-supplied acr_values would result in potentially allowing "lower" ACRs to satisfy the requirement, which would defeat the original purpose of actually enforcing some kind of minimum baseline.

An alternative seems to be looking up the LoA that is mapped to the ACR, and then filtering out any values that map to a "lower" LoA? This feels brittle and especially when mixing in custom LoA mappings at the client and realm level, it becomes even more confusing. If the client has an overridden map of LoA<->ACR, do we take into account *only* the client map? Do they somehow get merged if keys in one or the other are missing? I'm not quite sure what the behaviour here is even in the current implementation, without the proposed addition.

If a client application wants to (and is able to) request and verify a specific ACR, they are not really in need of this feature, at least in the scope I imagined: Enabling clients that *do not* have ACR capabilities to benefit from Keycloak's existing facilities. I'm open to discussion here, maybe you have use cases in mind that are outside of what we're considering on our end.

Q2) I think a realm-wide switch that can be overridden by clients is a good idea. I would suggest using the same attribute name. Placing it under "ACR to LoA Mapping" on the realm settings page seems to be the most reasonable place, as that is also where the current Default ACR Values configuration is present on clients. That way it's in "the same place" for both clients and the realm.

We might need to take some special care in the "overriding" scenario in this case: When the realm setting is set, and a client *does not* set its own value, that would mean the realm default applies. But what should we use as a "sentinel value" to denote that the client does not want the realm default and *also* does not want to specify its own minimum/force acr. That seems to be a decision that needs to be made regardless of whether the "minimal" or "force" strategy is chosen. I think an empty list (rather than null) would be reasonable here

Q3) I think in general the default_acr_values setting should be retained. It's part of [OIDC client registration](https://openid.net/specs/openid-connect-registration-1_0.html), and removing it be would be a regression. Handling cases where both are set is difficult:

If the "forced" strategy is chosen, I would suggest simply rejecting clients that set both as invalid configuration. This becomes a little bit more difficult if we also introduce a realm-level force attribute, but even in the case "realm forced setting is set" and "client sets default acr_values", I think the client should be rejected, unless the "force_acr_values" is set to an empty list, signaling that the client wants to opt out of the realm-wide default.

If the "minimal" strategy is chosen, we probably have to take into account the mapped LoA again, and check if any of the mapped LoA for any of the default_acr_values are *below* the minimum set, and reject the configuration as invalid based on that.

In general, I think rejecting config here as opposed to trying to make sense of seemingly confusing configuration is a benefit - it reduces the risk of accidentally misconfiguring the ACR/LoA settings and allowing in users with insecure login methods when we want to enforce a secure login

sonOfRa commented:
That plan sounds good to me, our needs are covered by implementing it this way. Before I get started rewriting this PR, I have some questions/points on how the implementation should behave:

1. We should use ONE minimum value, rather than allowing multiple minimum values

2. When a client explicitly requests an ACR that is *lower* than the minimum, I think the best path is to silently "upgrade" the request, simply ignoring that the client requested a lower ACR (as opposed to aborting the request with an error). That means in `getAcrValues`, when a client sets a minimum ACR value, we can either just return `List.of("minimumAcr")`, or alternatively, return a list that contains all the values the client requested, *minus* the ones that are lower than the minimum ACR, *plus* the minimum ACR, if it is not already present. Both ways should yield the desired result, do you have a preference here?

3. Rejecting "invalid" configuration. This seemed simple at first, but there are several scenarios to keep in mind here:
3.1. Updating a client is relatively straightforward, check either the realm ACR/LoA mapping or the one on the client if set, and reject the config if we can't find a LoA for the given minimum ACR value. When the LoA mapping is updated, we may also have to validate again and check if the minimum value (if present) is valid given the updated mapping
3.2 When a realm's ACR/LoA mapping is updated, we would have to enumerate all clients and evaluate the minimum values of all clients for consistency. This seems expensive, and I feel like we might not want to do this.
But if we don't actually validate the config when the realm is updated, we might also not want to validate when the client is updated, because it's more consistent that way?
If we don't do any validation, we also have to consider behaviour if the configured minimum ACR is not mapped. Probably the most reasonable behaviour is simply retaining the current behaviour: Run into `NumberFormatException` in `mapLoaToAcr` and use `-1` as the resulting LoA value and fall back to Keycloak's default behaviour

4. `OIDCClientRepresentation` handling. I added them to that class because the spec says "Additional Client Metadata parameters MAY also be used. Some are defined by other specifications, such as [OpenID Connect RP-Initiated Logout 1.0](https://openid.net/specs/openid-connect-registration-1_0.html#OpenID.RPInitiated).". I can remove them again if we don't want to expose Keycloak-specific settings here.

5. On avoiding realm level: I agree, not adding a realm wide flag makes things easier to implement, and a realm-wide implementation can easily be enforced by configuring the authentication flow to simply not providing a way to log in with a lower ACR than desired.

mposolda commented:
@sonOfRa Thanks for the discussion and for raising the follow-up clarifications! I am adding some answers inline, but overally, your ideas look nice to me.

> That plan sounds good to me, our needs are covered by implementing it this way. Before I get started rewriting this PR, I have some questions/points on how the implementation should behave:
> 
>     1. We should use ONE minimum value, rather than allowing multiple minimum values
+1

> 
>     2. When a client explicitly requests an ACR that is _lower_ than the minimum, I think the best path is to silently "upgrade" the request, simply ignoring that the client requested a lower ACR (as opposed to aborting the request with an error). That means in `getAcrValues`, when a client sets a minimum ACR value, we can either just return `List.of("minimumAcr")`, or alternatively, return a list that contains all the values the client requested, _minus_ the ones that are lower than the minimum ACR, _plus_ the minimum ACR, if it is not already present. Both ways should yield the desired result, do you have a preference here?

This is interesting concern. My vote is that:
- In case that ACR is requested by `claims` parameter as "essential", then we reject the request if the requested claims cannot meet the minimum for the client
- In case that ACR is requested by `claims` or `acr_values` parameter as not "essential", then +1 to silently "upgrade" the request. 

For the second case, I don't have much preference between `List.of("minimumAcr")` or removing lower ACRs from request than the minimum. Maybe I am slightly more for `List.of("minimumAcr")` as it is slightly less complex and error-prone :-)  But for both cases, I suggest to add some `logger.warn` with mention that some acr values provided as parameters were ignored.

> 
>     3. Rejecting "invalid" configuration. This seemed simple at first, but there are several scenarios to keep in mind here:
>        3.1. Updating a client is relatively straightforward, check either the realm ACR/LoA mapping or the one on the client if set, and reject the config if we can't find a LoA for the given minimum ACR value. When the LoA mapping is updated, we may also have to validate again and check if the minimum value (if present) is valid given the updated mapping
+1

>        3.2 When a realm's ACR/LoA mapping is updated, we would have to enumerate all clients and evaluate the minimum values of all clients for consistency. This seems expensive, and I feel like we might not want to do this.
+1 to not doing this expensive operation. I think we may already have this issue because if realm ACR-to-Loa mapping is updated, the client `default ACR values` may still have the original values, which may not be valid anymore.

>        But if we don't actually validate the config when the realm is updated, we might also not want to validate when the client is updated, because it's more consistent that way?
>        If we don't do any validation, we also have to consider behaviour if the configured minimum ACR is not mapped. Probably the most reasonable behaviour is simply retaining the current behaviour: Run into `NumberFormatException` in `mapLoaToAcr` and use `-1` as the resulting LoA value and fall back to Keycloak's default behaviour
+1 for the Keycloak default behaviour. ATM it is already done in a way that if non-existing claim is requested as "essential", we reject the request. If non-essential, we log the warning and continue https://github.com/keycloak/keycloak/blob/26.0.0/services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java#L329-L338 

> 
>     4. `OIDCClientRepresentation` handling. I added them to that class because the spec says "Additional Client Metadata parameters MAY also be used. Some are defined by other specifications, such as [OpenID Connect RP-Initiated Logout 1.0](https://openid.net/specs/openid-connect-registration-1_0.html#OpenID.RPInitiated).". I can remove them again if we don't want to expose Keycloak-specific settings here.
+1 for remove it from `OIDCClientRepresentation`. Right now, we use just parameters, which are present in either one of OIDC or OAuth2 specifications. 

> 
>     5. On avoiding realm level: I agree, not adding a realm wide flag makes things easier to implement, and a realm-wide implementation can easily be enforced by configuring the authentication flow to simply not providing a way to log in with a lower ACR than desired.

+1

sonOfRa committed:
Enable enforcement of a minimum ACR at the client level (#16884)

Signed-off-by: Simon Levermann <github@simon.slevermann.de>

sonOfRa commented:
Just pushed a starting point for the discussed changes. Currently writing some tests, but manual testing failed because the head of main branch currently won't start for me due to an exception:

```
❯ java -jar quarkus/server/target/lib/quarkus-run.jar start-dev --verbose
ERROR: Unexpected error when starting the server in (development) mode
Error details:
java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
        at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
        at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:300)
        at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:103)
        at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:200)
        at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
        at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:146)
        at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:57)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.keycloak.quarkus.runtime.cli.Picocli.run(Picocli.java:147)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:135)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:106)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
Caused by: java.lang.NoClassDefFoundError: oracle/sql/CharacterSet
        at io.quarkus.jdbc.oracle.runtime.OracleInitRecorder.setupCharSets(OracleInitRecorder.java:21)
        at io.quarkus.deployment.steps.ExtendedCharactersSupport$preinitializeCharacterSets1608602642.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.ExtendedCharactersSupport$preinitializeCharacterSets1608602642.deploy(Unknown Source)
        ... 27 more
Caused by: java.lang.ClassNotFoundException: oracle.sql.CharacterSet
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:114)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:72)
        ... 30 more
```

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/mappers/AcrProtocolMapper.java:
```diff
@@ -109,6 +109,7 @@ protected String getAcr(AuthenticatedClientSessionModel clientSession) {
 
         Map<String, Integer> acrLoaMap = AcrUtils.getAcrLoaMap(clientSession.getClient());
+        // Do we want to add the enforcement and produce an error here in the mapper as well?
         String acr = AcrUtils.mapLoaToAcr(loa, acrLoaMap, AcrUtils.getRequiredAcrValues(
```
@sonOfRa I think that responsibility of the `AcrProtocolMapper` should not be to do any additional checks, but just map the claim from the clientSession into the token.

The validations should be done in the layers (like for example `AuthorizationEndpoint`as you pointed) to make sure that authentication doesn't even start if the sent level in the `claims` is not correct and too low.

In the most ideal case, the validations should be done rather in the `AuthorizationEndpointChecker` rather than in `AuthorizationEndpoint.updateAuthenticationSession()` to make sure that request fail-fast even before authenticationSession is created (or also it will fail-fast for the PAR request). But I don't know if you want to update this point in your PR as ACR checks are already done before your PR in the `AuthorizationEndpoint.updateAuthenticationSession()`. So feel free to keep as it is in your PR (Just please remove the comment in `AcrProtocolMapper`)

sonOfRa replied:
I'm open to moving it to `AuthorizationEndpointChecker` but to me it looks like that at least for acr, the checks and the part that updates the authenticationSession is quite closely intertwined with the setting of the client notes (belongs in `updateAuthenticationSession()` and the actual checking of the ACR values (belongs in `AuthorizationEndpointChecker`). I have no particularly strong feelings here.

I'll remove the comment, it was just a marker for me where to add the in-line review comment.

mposolda replied:
@sonOfRa Yes, I see your point. I may take a look at this at some point, but not yet sure as it kind of works in a way it is done today (even if not 100% ideal as we may possibly handle it even before auth-session creation to possibly save some data etc). Thanks for removing the comment!

mposolda reviewed (commented):
@sonOfRa Thanks for the updates!

Added some minor comment inline about possibly removing the comment in `AcrProtocolMapper` (if you want to move checks to AuthorizationEndpointChecker, it is welcome, but just if you feel strongly that it should be done in this PR :-)

Besides that: Is it possible to add tests for:
- Scenario that authentication request fails when requested essential claim with lower value than the "minimum" level?
- Scenario when attempt to update client fails with the attempt to use the "minimum" value, which is not mapped to any level?

mposolda commented:
@thomasdarimont Do you want to review as well?

sonOfRa commented on js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties:
```diff
@@ -1358,6 +1358,7 @@ authorizationEncryptedResponseAlgHelp=JWA Algorithm used for key management in e
 deleteConfirmGroup_other=Are you sure you want to delete these groups.
 scopePermissions.users.manage-description=Policies that decide if an administrator can manage all users in the realm
 defaultACRValuesHelp=Default values to be used as voluntary ACR in case that there is no explicit ACR requested by 'claims' or 'acr_values' parameter in the OIDC request.
+minimumACRValueHelp=Minimum ACR to be enforced by keycloak. Overrides lower ACRs explicitly requested via 'acr_values' or 'claims', unless marked they are essential
```
It's a singular value, but I just realized in the UI part of the PR I added it as a MultiLine input. That was a copy-pasting accident from a previous version of this PR where it was multiple values. I'll adjust the input type.

sonOfRa commented:
> * Scenario that authentication request fails when requested essential claim with lower value than the "minimum" level?
> 
> * Scenario when attempt to update client fails with the attempt to use the "minimum" value, which is not mapped to any level?

I'll write some additional tests for those scenarios

sonOfRa commented:
Added the requested additional tests (and some more that seemed to make sense to me). Found a minor bug in the process where I only took into account numerical ACRs that directly match LoAs from authentication flows, which is fixed with the introduction of the `getLoaForAcr(...)` method.

Also changed the input to be a `TextControl` rather than `MultiLineInput` to reflect that it's a singular value, as well as added an updated screenshot of the advanced client configuration section from the admin UI to reflect the newly added field

mposolda reviewed (approved):
@sonOfRa Thanks for all the updates! I hope to merge if tests are OK

@agagancarczyk Also thanks to you for the review!

mposolda committed:
Merge branch 'main' into enforce-loa
